### PR TITLE
*: do not block etcdserver when encoding store into json

### PR DESCRIPTION
**Pull request by xiang90**
only blocks when doing the memory copy; do not block on encoding

### CODE REVIEW

File: etcdserver/server.go [hunk 1]
```diff
@@ -809,36 +809,39 @@ func (s *EtcdServer) applyConfChange(cc raftpb.ConfChange, confState *raftpb.Con
...
-	}
-	err = s.r.raftStorage.Compact(snapi, confState, d)
-	if err != nil {
-		// the snapshot was done asynchronously with the progress of raft.
-		// raft might have already got a newer snapshot and called compact.
-		if err == raft.ErrCompacted {
-			return
+	clone := s.store.Clone()
+
+	go func() {
```
Comment (yichengq) [inline etcdserver/server.go]: move the goroutine inside and pass the store clone in as an argument of the function?
I think it is easy for testing and readability.

File: etcdserver/server.go [hunk 2]
```diff
@@ -809,36 +809,39 @@ func (s *EtcdServer) applyConfChange(cc raftpb.ConfChange, confState *raftpb.Con
...
+		d, err := clone.SaveNoCopy()
+		// TODO: current store will never fail to do a snapshot
+		// what should we do if the store might fail?
+		if err != nil {
+			log.Panicf("etcdserver: store save should never fail: %v", err)
 		}
-		log.Panicf("etcdserver: unexpected compaction error %v", err)
-	}
-	log.Printf("etcdserver: compacted log at index %d", snapi)
+		err = s.r.raftStorage.Compact(snapi, confState, d)
```
Comment (barakmich) [inline etcdserver/server.go]: Double check: is it safe to operate on `s` at this point?

Comment (barakmich) [inline etcdserver/server.go]: Since we're doing this, why do we keep the whole array of bytes in memory instead of .Compact taking an `io.Reader`?

Comment (xiang90) [inline etcdserver/server.go]: @barakmich snapshot in raft is an array of bytes. attempting to change snapshot to io.reader is pre-mature until we "fix" raft snapshot.

Comment (xiang90) [inline etcdserver/server.go]: it actually only accesses raftStorage, which is thread-safe.

File: etcdserver/server.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/server.go]: i do not think we should separate the routine from the function. they are highly tight together at this point. we will revisit it later if we want to refactor the whole thing.

**Comment by barakmich** [comment]
LGTM

**Comment by xiang90** [comment]
@yichengq This is at your hand now. I now this would conflict with 210 branch if we merge here.

**Comment by xiang90** [comment]
@yichengq Can you move this to 210 branch? Or do you want me to move it?

**Comment by yichengq** [comment]
welcome to move it. i plan to merge it when i rebase 210 on master.

**Comment by xiang90** [comment]
ok. will do

**Comment by xiang90** [comment]
closed in favor of https://github.com/coreos/etcd/pull/2386

**Pull request closed without merging**